File: cpp/se/ArDomainUnrenewCommand.cpp
#include "se/ArDomainUnrenewCommand.hpp"

#include "se/CommandType.hpp"
#include "se/ArExtension.hpp"
#include "se/ArDomainObjectType.hpp"
#include "se/EPPDateFormatter.hpp"
#include "xml/XMLHelper.hpp"
#include "se/XMLGregorianCalendar.hpp"

namespace {
    class ArUnrenewCommand : public CommandType
    {
        public:
            ArUnrenewCommand() : CommandType (getCommandName()) { }
            std::string getCommandName() const { return "unrenew"; }
            std::string toString() const { return "unrenew"; }
    };

    Extension& arExtension() {
        static Extension* arExt = new ArExtension();
        return *arExt;
    }

    const ArUnrenewCommand unrenewCmdType;
    const ArDomainObjectType ardomType;
}; // anonymous namespace

ArDomainUnrenewCommand::ArDomainUnrenewCommand(
        const std::string &name,
        const XMLGregorianCalendar& exDate) : ProtocolExtensionCommand(
            &unrenewCmdType, &ardomType, name, arExtension())
{
    std::string exDateStr = EPPDateFormatter::toXSDate(exDate);
    XMLHelper::setTextContent
        (xmlWriter->appendChild(objElement, "curExpDate"), exDateStr);
}

File: src/main/java/com/ausregistry/jtoolkit2/se/ArDomainUnrenewCommand.java
package com.ausregistry.jtoolkit2.se;

import java.util.GregorianCalendar;

import com.ausregistry.jtoolkit2.EPPDateFormatter;
import com.ausregistry.jtoolkit2.ErrorPkg;

/**
 * Mapping of EPP urn:ar:params:xml:ns:arext-1.0 domainUnrenew command
 * specified by the AusRegistry EPP extensions document.
 * Use this class to generate an AusRegistry-compliant XML document, given
 * simple input parameters.  The toXML method in Command serialises this object
 * to XML.
 *
 * @see ArDomainUnrenewResponse
 */
public class ArDomainUnrenewCommand extends ProtocolExtensionCommand {
    private static final long serialVersionUID = -3723213074751854975L;

    private static final CommandType UNRENEW_COMMAND_TYPE = new ArUnrenewCommandType();

    /**
     * @throws IllegalArgumentException if {@code name} or {@code exDaten} is {@code null}.
     */
    public ArDomainUnrenewCommand(String name, GregorianCalendar exDate) {
        super(UNRENEW_COMMAND_TYPE, ExtendedObjectType.AR_DOMAIN, name);

        if (exDate == null) {
            throw new IllegalArgumentException(
                    ErrorPkg.getMessage("se.ar.unrenew.missing_arg"));
        }

        String exDateStr = EPPDateFormatter.toXSDate(exDate);
        xmlWriter.appendChild(objElement, "curExpDate").setTextContent(
                exDateStr);
    }

    @Override
    protected Extension getExtension() {
        return ExtensionImpl.AR;
    }
}

